Add Governor documentation (#1235)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/api/governance.adoc

:VotingUnitsTrait: xref:VotingUnitsTrait[VotingUnitsTrait]
:VotesComponent: xref:VotesComponent[VotesComponent]
:IVotes: xref:IVotes[IVotes]
:governor: xref:governance/governor.adoc[Governor]
:inner-src5: xref:api/introspection.adoc#ISRC5[SRC5 ID]
:GovernorComponent: xref:#GovernorComponent[GovernorComponent]
:TimelockControllerComponent: xref:#TimelockControllerComponent[TimelockControllerComponent]

= Governance

include::../utils/_common.adoc[]

This crate includes primitives for on-chain governance.

== Governor

This modular system of Governor components allows the deployment of easily customizable on-chain voting protocols.

TIP: For a walkthrough of how to implement a Governor, check the {governor} page.

[.contract]
[[IGovernor]]
=== `++IGovernor++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.20.0-rc.0/packages/governance/src/governor/interface.cairo[{github-icon},role=heading-link]

[.hljs-theme-dark]
```cairo
use openzeppelin_governance::governor::interface::IGovernor;
```

Interface of a governor contract.

[.contract-index]
.{inner-src5}
--

> 0x1

Author: True, I will put the right number today.

Must return the minimum number of votes that an account must have to create a proposal.

[.contract-item]
[[GovernorComponent-GovernorQuorumTrait-quorum]]
==== `[.contract-item-name]#++quorum++#++(self: @ContractState, timepoint: u64) → u256++` [.item-kind]#extension#

Must return the minimum number of votes required for a proposal to succeed.

[.contract-item]
[[GovernorComponent-GovernorCountingTrait-counting_mode]]
==== `[.contract-item-name]#++counting_mode++#++(self: @ContractState) → ByteArray++` [.item-kind]#extension#

Must return a description of the possible `support` values for `cast_vote` and the way these votes are counted,
meant to be consumed by UIs to show correct vote options and interpret the results. See `COUNTING_MODE()` for more details.

[.contract-item]
[[GovernorComponent-GovernorCountingTrait-count_vote]]
==== `[.contract-item-name]#++count_vote++#++(ref self: ContractState, proposal_id: felt252, account: ContractAddress, support: u8, total_weight: u256, params: Span<felt252>) → u256++` [.item-kind]#extension#

Must register a vote for `proposal_id` by `account` with a given `support`, voting `weight` and voting `params`.

> 

Author: With this said, we should always backtick parameters in these descriptions with maybe this exception, I just don't think is worth checking and refactoring every description block to mention and backtick all the parameters.

> 

Reviewer: Fair points. If I find any instances where backticked params would without a doubt improve the description, I'll highlight them. Otherwise, keeping them as is works for me 👍

functions that are restricted with the `assert_only_governance` guard, and also
potentially execute transactions on behalf of the governor. Because of this, this module
is designed to work with the TimelockController as the unique potential external
executor.

[.contract-item]
[[GovernorComponent-GovernorExecutionTrait-execute_operations]]
==== `[.contract-item-name]#++execute_operations++#++(ref self: ContractState, proposal_id: felt252, calls: Span<Call>)++` [.item-kind]#internal#

Execution mechanism. Can be used to modify the way execution is
performed (for example adding a vault/timelock).

[.contract-item]
[[GovernorComponent-GovernorExecutionTrait-queue_operations]]
==== `[.contract-item-name]#++queue_operations++#++(ref self: ContractState, proposal_id: felt252, calls: Span<Call>)++` [.item-kind]#internal#

Queuing mechanism. Can be used to modify the way queuing is
performed (for example adding a vault/timelock).

Requirements:

- Must return a timestamp that describes the expected ETA for execution. If the returned
value is 0, the core will consider queueing did not succeed, and the public `queue`
function will revert.

[.contract-item]

> [[GovernorComponent-GovernorExecutionTrait-proposal_needs_queueing]]
> ==== `[.contract-item-name]#++proposal_needs_queueing++#++(self: @ContractState) → bool++` [.item-kind]#internal#

Reviewer: ```suggestion
[[GovernorComponent-GovernorExecutionTrait-proposal_needs_queuing]]
==== `[.contract-item-name]#++proposal_needs_queuing++#++(self: @ContractState) → bool++` [.item-kind]#internal#
```

- `params=fractional` might refer to a scheme where votes are divided fractionally between
for/against/abstain.
- `params=erc721` might refer to a scheme where specific NFTs are delegated to vote.

NOTE: The string can be decoded by the standard https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams[`URLSearchParams`]
JavaScript class.

[.contract-item]
[[GovernorComponent-hash_proposal]]
==== `[.contract-item-name]#++hash_proposal++#++(calls: Span<Call>, description_hash: felt252) → felt252++` [.item-kind]#external#

Hashing function used to (re)build the proposal id from the proposal details.

[.contract-item]
[[GovernorComponent-state]]
==== `[.contract-item-name]#++state++#++(proposal_id: felt252) → ProposalState++` [.item-kind]#external#

Returns the state of a proposal, given its id.

[.contract-item]
[[GovernorComponent-proposal_threshold]]
==== `[.contract-item-name]#++proposal_threshold++#++() → u256++` [.item-kind]#external#

The number of votes required in order for a voter to become a proposer.

[.contract-item]
[[GovernorComponent-proposal_snapshot]]
==== `[.contract-item-name]#++proposal_snapshot++#++(proposal_id: felt252) → u64++` [.item-kind]#external#

> Timepoint used to retrieve user's votes and quorum. If using block number, the snapshot is performed at the

Author: It is the timepoint that will be used every time votes are queried for both users and quorum for this proposal. Not sure how to rephrase it, any suggestion is welcomed. Maybe this sentence?

> Timepoint used to retrieve user's votes and quorum. If using block number, the snapshot is performed at the

Reviewer: Ahhh my bad, I see. The `to` makes me think the timepoint is doing the retrieving. Perhaps something like this:

```suggestion
Timepoint used when retrieving user's votes and quorum. If using block number, the snapshot is performed at the
```

==== `[.contract-item-name]#++cast_vote_with_reason++#++(proposal_id: felt252, support: u8, reason: ByteArray) → u256++` [.item-kind]#external#

Cast a vote with a `reason`.

Requirements:

- The proposal must be active.

Emits a {VoteCast} event.

[.contract-item]
[[GovernorComponent-cast_vote_with_reason_and_params]]
==== `[.contract-item-name]#++cast_vote_with_reason_and_params++#++(proposal_id: felt252, support: u8, reason: ByteArray, params: Span<felt252>) → u256++` [.item-kind]#external#

Cast a vote with a `reason` and additional serialized `params`.

Requirements:

- The proposal must be active.

Emits either:

- {VoteCast} event if no params are provided.
- {VoteCastWithParams} event otherwise.

[.contract-item]
[[GovernorComponent-cast_vote_by_sig]]
==== `[.contract-item-name]#++cast_vote_by_sig++#++(proposal_id: felt252, support: u8, voter: ContractAddress, signature: Span<felt252>) → u256++` [.item-kind]#external#

Cast a vote using the `voter`'s signature.

Requirements:

- The proposal must be active.
- The nonce in the signed message must match the account's current nonce.
- `voter` must implement `SRC6::is_valid_signature`.

> - `signature` should be valid for the message hash.

Author: It must be must.

Emits a {VoteCast} event.

[.contract-item]
[[GovernorComponent-cast_vote_with_reason_and_params_by_sig]]
==== `[.contract-item-name]#++cast_vote_with_reason_and_params_by_sig++#++(proposal_id: felt252, support: u8, voter: ContractAddress, reason: ByteArray, params: Span<felt252>, signature: Span<felt252>) → u256++` [.item-kind]#external#

Cast a vote with a `reason` and additional serialized `params` using the `voter`'s
signature.

Requirements:

- The proposal must be active.
- The nonce in the signed message must match the account's current nonce.
- `voter` must implement `SRC6::is_valid_signature`.

> - `signature` should be valid for the message hash.

Author: Here it should be must 😺

Emits either:

- {VoteCast} event if no params are provided.
- {VoteCastWithParams} event otherwise.

[.contract-item]
[[GovernorComponent-nonces]]
==== `[.contract-item-name]#++nonces++#++(voter: ContractAddress) → felt252++` [.item-kind]#external#

Returns the next unused nonce for an address.

[.contract-item]
[[GovernorComponent-relay]]
==== `[.contract-item-name]#++relay++#++(call: Call)++` [.item-kind]#external#

Relays a transaction or function call to an arbitrary target.

In cases where the governance executor is some contract other than the governor itself, like
when using a timelock, this function can be invoked in a governance proposal to recover
tokens that were sent to the governor contract by mistake.

NOTE: If the executor is simply the governor itself, use of `relay` is redundant.

[#GovernorComponent-Internal-Functions]
==== Internal functions

[.contract-item]
[[GovernorComponent-initializer]]
==== `[.contract-item-name]#++initializer++#++(ref self: ContractState)++` [.item-kind]#internal#

> Initializes the contract by registering the supported interface id.

Author: We should add the right Id in the right place instead, I will update it today.

==== `[.contract-item-name]#++current_quorum_numerator++#++(self: @ContractState) → u256++` [.item-kind]#external#

Returns the current quorum numerator value.

[.contract-item]
[[GovernorVotesQuorumFractionComponent-quorum_numerator]]
==== `[.contract-item-name]#++quorum_numerator++#++(self: @ContractState, timepoint: u64) → u256++` [.item-kind]#external#

Returns the quorum numerator value at a specific `timepoint` in the past.

[.contract-item]
[[GovernorVotesQuorumFractionComponent-quorum_denominator]]
==== `[.contract-item-name]#++quorum_denominator++#++(self: @ContractState) → u256++` [.item-kind]#external#

Returns the quorum denominator value.

[#GovernorVotesQuorumFractionComponent-Internal-Functions]
==== Internal functions

[.contract-item]
[[GovernorVotesQuorumFractionComponent-initializer]]
==== `[.contract-item-name]#++initializer++#++(self: @ComponentState<TContractState>, votes_token: ContractAddress, quorum_numerator: u256)++` [.item-kind]#internal#

Initializes the component by setting the votes token and the initial quorum numerator value.

Requirements:

* `votes_token` must not be zero.
* `quorum_numerator` must be less than `quorum_denominator`.

> Emits a {QuorumNumeratorUpdated} event.

Reviewer: Not sure why, but this isn't displaying the link

[.contract-item]
[[GovernorVotesQuorumFractionComponent-update_quorum_numerator]]
==== `[.contract-item-name]#++update_quorum_numerator++#++(self: @ComponentState<TContractState>, new_quorum_numerator: u256)++` [.item-kind]#internal#

Updates the quorum numerator.

NOTE: This function does not emit an event if the new quorum numerator is the same as the old one.

Requirements:

* `new_quorum_numerator` must be less than `quorum_denominator`.

> May emit a {QuorumNumeratorUpdated} event.

Reviewer: Same link issue here

[.contract-item]
[[GovernorTimelockExecutionComponent-update_timelock]]
==== `[.contract-item-name]#++update_timelock++#++(ref self: ContractState, new_timelock: ContractAddress)++` [.item-kind]#external#

Updates the associated timelock.

Requirements:

- The caller must be the governance.

Emits a {TimelockUpdated} event.

[#GovernorTimelockExecutionComponent-Internal-Functions]
==== Internal functions

[.contract-item]
[[GovernorTimelockExecutionComponent-initializer]]
==== `[.contract-item-name]#++initializer++#++(ref self: ContractState, timelock: ContractAddress)++` [.item-kind]#internal#

Initializes the timelock controller.

Requirements:

- The timelock must not be the zero address.

[.contract-item]
[[GovernorTimelockExecutionComponent-assert_only_governance]]
==== `[.contract-item-name]#++assert_only_governance++#++(self: @ContractState)++` [.item-kind]#internal#

Ensures the caller is the executor (the timelock controller in this case).

[.contract-item]
[[GovernorTimelockExecutionComponent-timelock_salt]]
==== `[.contract-item-name]#++timelock_salt++#++(self: @ContractState, description_hash: felt252) → felt252++` [.item-kind]#internal#

> Computes the `TimelockController` operation salt.

Reviewer: ```suggestion
Computes the `TimelockController` operation salt as the XOR of the governor address and `description_hash`.
```
In case users are curious  🤷‍♂️
